Typography [fixes #763]

carlfairclough opened the pull request:
Updated all typography

## Description

Moved almost all typography rules to live in `T.vue` which can either be called with `<T props />` or with classes (unscoped). It follows tailwind/tachyons/atomic/etc style css rules with low levels of specificity. This is a highly rigid approach but enables us to ensure that type goes through a single funnel, and minimizes custom styles.

- Default type color classes have been defined which will switch on darkmode without having to redefine them.

- Markdown content is styled with a chain of `:not()`s which prevent the styles from overriding any explicitly set. For example, `<div class="featured" />` has been replaced in markdown with the `markdown-it-attrs` package, allowing you to pass something such as
`# Lorem ipsum si dolor et amet {.l3}`, which will create a h1 element with `level 3` styles applied.  By default, `h1` uses `level 1``.l1`

- Some pages with repeated content now use `v-for` and and object array in `data:` to allow styles for the content blocks to be defined **once**. `/build/` is the most significant example of this.

## Related Issue

#763

samajammin commented on docs/dapps/index.md:
```diff
@@ -12,11 +12,7 @@ lang: en-US
-**Want to get started using Ethereum?** There are hundreds of decentralized applications (dapps) built on Ethereum that you can use today.
-
-</div>
+**Want to get started using Ethereum?** There are hundreds of decentralized applications (dapps) built on Ethereum that you can use today. {.l4 .featured}
```
This change will require updating all translation markdown pages as well. Grep for `class="featured"` & you'll see all the files (there's ~70).

carlfairclough replied:
Yeah I spotted, this isn't a breaking change right now though. Happy to pick that up when continuing

carlfairclough commented:
@samajammin I totally understand where you're coming from, it *is* a lot to get to grips with initially. But imo this has a number of benefits, mainly:

**No inheritance**
No type element should have text size set on the parent (or be overriding a text size. It means that we can see in the codebase, in the individual page template, the rules being applied. When somebody is just using markdown, the styles map like so:
- h1: l1
- h2: l3
- ...
- h6: l6
- p: l7
- ul li: l7

I think a reason it looks intimidating is the chain of `:not()`s, which was the fastest way to ensure that any levels set take precedence over any html tag (`# title text` is l1, however `# title text {.l3} would be level 3)

**High visibility & works according constraints**
Style rules don't need to be abstracted away to a `style` tag when using this, and *especially* not away to a collection of styles (e.g. theme.styl, dark.styl, etc.). It means pages/components only need to be edited in one place. The props & classes also act as guard rails to help people leverage and work with the system.

Now there is definitely a downside with atomic classes:
- They're almost a return to inline styles, especially when very clear class names are used (almost 1:1 with the styles)
- There is a learning curve when shortened class names are used (like the ones here)
- it can get messy when there are long chains of classes

The naming could be much clearer, but it is a 1:1 representation of the Figma file, allowing people to leverage the exact colours and type sizes (and max-widths) without having to apply create any extra classes. In most pages, like `Build`, the component + props are used. However, in Markdown, I've used {.classnames} which achieve the same thing. The component can be used there too though!

It's a little extra work, but I think all of this could be documented very easily — I could fork the repo to have a single page which shows `<T` (or a progression of that naming), alongside every `l1 l2 l3 l4 l5 l6 l7 l8` prop/class (`level1` instead?) and do the same for text colors, background colors, margins, and so on.

samajammin commented on docs/.vuepress/components/T.vue:
```diff
@@ -0,0 +1,342 @@
+      )
+    },
+
+    getModifiers() {
```
Seems these props could be replaced by just passing classnames, e.g. 
```<T p class="subtitle mw55ch s300 p l4 c-text">```
instead of:
```<T p l4 cText s300 class="subtitle mw55ch">```
since `getModifiers()` just adds those classes.

samajammin commented on docs/.vuepress/components/T.vue:
```diff
@@ -0,0 +1,342 @@
+        'p'
+      )
+    },
+    getLevel() {
```
Same here, we could just use classnames instead of passing props, e.g.
```<T p class="subtitle l4">```
instead of:
```<T p l4 class="subtitle">```

samajammin commented:
One change I noticed to be aware we account for:
![Image 2020-03-09 at 2 53 36 PM](https://user-images.githubusercontent.com/8097623/76260913-163b2980-6216-11ea-85bf-580b30a85a6e.png)
Looks like H1s now render an anchor link, which we want to avoid (i.e. keep the current format)

samajammin commented:
More changes to be aware of, looking at site preview:
![Image 2020-03-09 at 3 05 58 PM](https://user-images.githubusercontent.com/8097623/76261582-8bf3c500-6217-11ea-84ff-aa9bfc354fd2.png)
Spacing of the H2s looks tight.

samajammin commented:
More changes to be aware of, looking at site preview:
![Image 2020-03-09 at 3 08 04 PM](https://user-images.githubusercontent.com/8097623/76261705-d117f700-6217-11ea-907c-f5d644d2d242.png)
Headers are now covered by nav when scrolling to anchor text.

samajammin commented:
Another change on /languages page, card rows contain 2 vs 3 items:
![Image 2020-03-09 at 3 11 26 PM](https://user-images.githubusercontent.com/8097623/76261912-4daad580-6218-11ea-953e-c71ee1708c17.png)

carlfairclough commented:
All issues have been addressed, however some fairly significant changes have been introduced, the most major structurally, HTML/CSS wise, around handling content blocks *within* markdown.

### Anchors & headers
> h1s now render an anchor link
These have been removed — I thought they made sense given they are accessible via the sidebar. 
>Headers are now covered by nav when scrolling to anchor text.
The targets now have psuedo-elements which prevent them from being hidden behind the nav when automatically scrolled to:
```
  &:target:before {
    content: "";
    display: block;
    height: 60px; /* fixed header height*/
    margin: -60px 0 0; /* negative fixed header height */
  }
```
This brought in a secondary issue, where the sidebar links would ignore the new psuedo-target. I have addressed the issue, however it involved an update to how the sidebar links work — instead of using `router-link` there is now an `a` instead, and I cut all unnecessary styles in the process. 
I've applied the new text styles and stripped all custom css.

> Spacing of the H2s looks tight
Updated this so it can _breathe_. We should also update figma to reflect the change

### .md content column

> Another change on /languages page, card rows contain 2 vs 3 items
Currently, a max-width/width is defined on the content column, with components being called inside that (which is the case inside the languages page). To avoid md components from being constrained this way, a `max-width` has been applied as a base style to *all* text elements. I imagine there may be niggles in the future (e.g. using non-text elements in .md), however right now everything appears good and working, and it plays nicely with `.has-sidebar` pages.

### T.vue

Gone! But the utility classes are now heavily used

### Utility class generation

There is a helper function to avoid repetitive class creation, it can be used in either of the following ways:

```
styles = {
   '100': 25px,
   '200': 50px,
   '300': 50vw,
}

generateClasses('classname', '-custom, 'height', styles)
/// OUTPUT
.classname-100-custom { height: 25px }
.classname-200-custom { height: 50px }
.classname-300-custom { height: 50vw }

```
stylesAlt = {
   unit: 'em'
   values: {
     '100': 25,
     '200': 50,
     '300': 75,
   }
}

generateClasses('classname-', null, 'max-width', styles)
/// OUTPUT
.classname-100 { max-width: 25em }
.classname-200 { max-width: 50em }
.classname-300 { max-width: 50em }
```

samajammin commented on docs/.vuepress/components/BuildPage.vue:
```diff
@@ -215,132 +126,126 @@ export default {
+    return {
+      templates: [
+        {
+          title: this.translateString('page-build-hello-world-title'),
```
Same issue here as in HomePage.vue, you'll need to bring this into the template.

carlfairclough replied:
Resolved by moving this `computed:`

samajammin commented on docs/.vuepress/theme/styles/theme.styl:
```diff
@@ -277,5 +145,78 @@ table
+    padding-left .5em
+    margin-bottom .5em
+
+    &:before
```
This will create an issue for right-to-left languages, e.g. see https://deploy-preview-801--ethereumorg.netlify.com/ar/ vs. https://ethereum.org/ar/

samajammin replied:
Perhaps you can solve by overwriting when the `.right-to-left-text` class is present

carlfairclough replied:
This has been resolved with a relatively significant update, all utility classes, list style updates, and the homepage arrows now have x-padding/margin values contained like so:
```
.div:not(.rtl)
  .class
    padding-left: 1rem

.rtl
  .class
    padding-right 1em
```
I changed `.right-to-left-text` to `.rtl`

carlfairclough commented on docs/.vuepress/components/HomePage.vue:
```diff
@@ -1,240 +1,230 @@
+      introBlocks: [
+        {
+          title: this.translateString('page-home-section-individuals-title'),
+          display: this.contentVersion() >= 1.1,
```
Resolved by moving this `computed:`

carlfairclough commented on docs/.vuepress/components/HomePage.vue:
```diff
@@ -1,240 +1,230 @@
+    return {
+      introBlocks: [
+        {
+          title: this.translateString('page-home-section-individuals-title'),
```
Resolved by moving this `computed:`

carlfairclough commented on docs/.vuepress/theme/styles/theme.styl:
```diff
@@ -277,5 +145,78 @@ table
+.dark-mode p
+  @extend .dark-mode .tc-text300
+
+.content__default
```
yeah, this is targeting `.md` generated content

samajammin replied:
Cool - let's add a comment there explaining that. It's a bit confusing since grepping for that class across the codebase yields no results.

carlfairclough commented on docs/.vuepress/theme/styles/utils.styl:
```diff
@@ -13,8 +42,292 @@
+
+// We then generate the classes for the default (.tc-{name}{shade})
+// and for :hover (.tc-h-{name}{shade})
+generateClasses('tc-text', null, 'color', $textColorObj)
```
tc here is `(text) color`, I imagine it growing to include background-color `bgc`. I haven't seen tc used as that before, but maybe `t-c` `bg-c` would be better? I'm keen to keep these as short as possible

samajammin replied:
K. We can keep as `tc`for now.

carlfairclough commented on docs/.vuepress/components/HomePage.vue:
```diff
@@ -1,240 +1,230 @@
-      </div>
+    </header>
 
+    <section class="intro-blocks flex flex-wrap pt-5">
```
Updated
![image](https://user-images.githubusercontent.com/4670881/76671573-cc37a800-6575-11ea-872c-802fb8a62d2e.png)
![image](https://user-images.githubusercontent.com/4670881/76671591-e4a7c280-6575-11ea-82de-276b3121707b.png)

carlfairclough commented on docs/.vuepress/theme/styles/theme.styl:
```diff
@@ -5,6 +5,9 @@
 *
   font-variant-ligatures none
 
+*, *:before, *:after
```
this is a global reset of the css box model and I more often see it used than not. Previously, there was a smattering of `box-sizing: border-box` and this reset ensures that every element uses `border-box`.

The default box model adds padding to the width of elements, which causes issues, for example `width:100vw; padding: 1em` will overflow the viewport.

samajammin commented:
Homepage is still broken for some translations, e.g. https://deploy-preview-801--ethereumorg.netlify.com/se/

samajammin commented on docs/.vuepress/theme/styles/theme.styl:
```diff
@@ -277,5 +145,78 @@ table
 
+// BASE TEXT STYLES
+// These target mainly our markdown content, and whenever a .l# class is not used
+// They should eventually be moved out of utils.styl
```
@carlfairclough

samajammin commented on docs/.vuepress/theme/styles/theme.styl:
```diff
@@ -277,5 +145,78 @@ table
+// They should eventually be moved out of utils.styl
+
+h1, h2, h3, h4, h5, h6, p, span, li {
+  max-width: $contentWidth;
```
Thoughts @carlfairclough?

samajammin commented on docs/.vuepress/theme/styles/utils.styl:
```diff
@@ -13,8 +42,292 @@
+
+// Set text colors
+// these should be revisited
+// alongside colors in config.vue
```
Thoughts @carlfairclough ?

samajammin commented:
Homepage highlight is broken for right to left languages, e.g. https://deploy-preview-801--ethereumorg.netlify.com/ar/
